contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,275 @@
+// SPDX-FileCopyrightText: 2022 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.9;
+
+interface IRestakingSink {
+    function receiveRestake() external payable;
+}
+
+/**
+ * @title A dedicated contract for handling stETH withdrawal request queue
+ * @author folkyatina
+ */
+contract WithdrawalQueue {
+    /**
+     * @notice minimal possible sum that is possible to withdraw
+     * We don't want to deal with small amounts because there is a gas spent on oracle 
+     * for each request. 
+     * But exact threshhold should be defined later when it will be clear how much will 
+     * it cost to withdraw.
+     */
+    uint256 public constant MIN_WITHDRAWAL = 0.1 ether;
+
+    /**
+     * @notice All state-modifying calls are allowed only from owner protocol. 
+     * @dev should be Lido
+     */ 
+    address payable public immutable OWNER;
+
+    /**
+     * @notice amount of ETH on this contract balance that is locked for withdrawal and waiting for claim
+     * @dev Invariant: `lockedEtherAmount <= this.balance`
+     */
+    uint256 public lockedEtherAmount = 0;
+
+    /// @notice queue for withdrawal requests
+    Request[] public queue;
+    
+    /// @notice length of the finalized part of the queue
+    uint256 public finalizedQueueLength = 0;
+
+    /// @notice structure representing a request for withdrawal.
+    struct Request {
+        /// @notice payable address of the recipient withdrawal will be transfered to
+        address payable recipient;
+        /// @notice block.number when the request created
+        uint96 requestBlockNumber;
+        /// @notice sum of the all requested ether including this request
+        uint256 cumulativeEther;
+        /// @notice sum of the all shares locked for withdrawal including this request
+        uint256 cumulativeShares;
+        /// @notice flag if the request was already claimed
+        bool claimed;
+    }
+
+    /// @notice finalization price history registry
+    Price[] public finalizationPrices;
+
+    /**
+     * @notice structure representing share price for some range in request queue
+     * @dev price is stored as a pair of value that should be devided later
+     */ 
+    struct Price {
+        uint128 totalPooledEther;
+        uint128 totalShares;
+        /// @notice last index in queue this price is actual for
+        uint256 index;
```

Comment:
can be `uint64`

Reply:
array length is uint256 by default. So, I prefer to use one type across the code for this until it saves us some gas.